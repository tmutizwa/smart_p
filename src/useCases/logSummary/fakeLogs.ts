export const fakeLog = `

/help_page/1 126.318.035.038
/contact 184.123.665.067

/home 184.123.665.067
/about/2 444.701.448.104
/help_page/1 646.865.545.408
/home 235.313.352.950
/contact 184.123.665.067
/contact 184.123.665.067
/help_page/1 543.910.244.929
/help_page/1 543.910.244.929

`;

export const parsedFakeLog = {
  '/help_page/1': [
    '126.318.035.038',
    '646.865.545.408',
    '543.910.244.929',
    '543.910.244.929',
  ],
  '/contact': ['184.123.665.067', '184.123.665.067', '184.123.665.067'],
  '/home': ['184.123.665.067', '235.313.352.950'],
  '/about/2': ['444.701.448.104'],
};
