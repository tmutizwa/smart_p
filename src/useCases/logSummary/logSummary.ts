import { promises as fs } from 'fs';
import path from 'path';

import { ReportType, PageViewLogs, PageViewTotals } from '~/types';

export async function getLogsSummary(type?: ReportType) {
  try {
    const data = await fs.readFile(
      path.join(__dirname, 'webserver.log'),
      'utf8'
    );

    const rawLogs = getRawLogs(data);

    return type === 'totalViews'
      ? getPageTotalViewsSummary(rawLogs)
      : getUniqueViewsummary(rawLogs);
  } catch (e) {
    throw 'Error reading log file, ' + e;
  }
}

function getUniqueViewsummary(rawLogs: PageViewLogs) {
  const viewsSummary: PageViewTotals = {};

  for (const key in rawLogs) {
    viewsSummary[key] = new Set(rawLogs[key]).size;
  }

  return viewsSummary;
}

function getPageTotalViewsSummary(rawLogs: PageViewLogs) {
  const views: PageViewTotals = {};

  for (const key in rawLogs) {
    views[key] = rawLogs[key].length;
  }

  return views;
}

export function getRawLogs(data: string): PageViewLogs {
  const rawLog: PageViewLogs = {};
  data
    .split('\n')
    .filter((line: string) => line.trim() !== '')
    .forEach((line: string) => {
      const [page, ip] = line.split(' ');

      if (!rawLog[page]) {
        rawLog[page] = [ip];
      } else {
        rawLog[page].push(ip);
      }
    });

  return rawLog;
}
