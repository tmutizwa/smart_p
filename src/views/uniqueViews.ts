import chalk from 'chalk';
import boxen from 'boxen';

import { PageViewTotals } from '~/types';

export function displayUniqueViews(views: PageViewTotals) {
  type sortableLog = [string, number];
  const orderedSummary: sortableLog[] = [];
  let summary = '';

  for (const page in views) {
    orderedSummary.push([page, views[page]]);
  }
  orderedSummary.sort((a, b) => b[1] - a[1]);

  orderedSummary.forEach((rec) => {
    summary += `${chalk.greenBright(rec[0])} ${chalk.red(
      rec[1]
    )} unique views \n`;
  });

  console.log(chalk.underline('Unique Views'));
  console.log(
    boxen(summary, {
      padding: 1,
      margin: 1,
      borderStyle: 'round',
      borderColor: 'green',
    })
  );
}
