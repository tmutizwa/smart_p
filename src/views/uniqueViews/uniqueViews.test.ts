import { displayUniqueViews } from './uniqueViews';
import { fakeUniqueViewsSummary } from '../../useCases/logSummary/fakeLogs';

describe('Unique Views display', () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should log the data correctly', () => {
    const consoleMock = jest.fn();
    jest.spyOn(console, 'log').mockImplementation(consoleMock);

    displayUniqueViews(fakeUniqueViewsSummary);

    expect(consoleMock.mock.calls[0][0]).toMatchSnapshot();
    expect(consoleMock.mock.calls[1][0]).toMatchSnapshot();
  });
});
