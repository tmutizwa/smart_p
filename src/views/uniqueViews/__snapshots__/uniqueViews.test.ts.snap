// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unique Views display should log the data correctly 1`] = `"[4mUnique Views[24m"`;

exports[`Unique Views display should log the data correctly 2`] = `
"[32m[39m
[32m   â•­â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®[39m
   [32mâ”‚[39m                                  [32mâ”‚[39m
   [32mâ”‚[39m   [92m/help_page/1[39m [31m3[39m unique views    [32mâ”‚[39m
   [32mâ”‚[39m   [92m/home[39m [31m2[39m unique views           [32mâ”‚[39m
   [32mâ”‚[39m   [92m/contact[39m [31m1[39m unique views        [32mâ”‚[39m
   [32mâ”‚[39m   [92m/about/2[39m [31m1[39m unique views        [32mâ”‚[39m
   [32mâ”‚[39m                                  [32mâ”‚[39m
   [32mâ”‚[39m                                  [32mâ”‚[39m
[32m   â•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯[39m
[32m[39m"
`;
